src/main/scala/io/iohk/ethereum/domain/BlockBody.scala

```diff
@@ -3,19 +3,136 @@ package io.iohk.ethereum.domain
 import io.iohk.ethereum.domain.BlockHeaderImplicits._
 import io.iohk.ethereum.rlp.{RLPEncodeable, RLPList, RLPSerializable, rawDecode}
 
-case class BlockBody(transactionList: Seq[SignedTransaction], uncleNodesList: Seq[BlockHeader]) {
-  override def toString: String =
-    s"BlockBody{ transactionList: $transactionList, uncleNodesList: $uncleNodesList }"
+import scala.collection.immutable
+
+class BlockBody private(
+                        private val transactionList: Array[SignedTransaction],
+                        val uncleNodesList: Seq[BlockHeader]
+                      ) extends Iterable[SignedTransaction] {
+
```

Comment:
As I mentioned on our call, I think you can get the benefit of O(1) access for indexing, without having to override `hashCode` and `equals` by using `IndexedSeq`:

```scala
case class BlockBody(transactions: IndexedSeq[SignedTransaction], uncleNodes: Seq[BlockHeader])
```

If you make sure that the case class is instantiated with an `Array` argument, it will be wrapped in a way that prevents mutability as well:
```console
scala> case class Test(xs: IndexedSeq[Int])
defined class Test

scala> val a = Test(Array(1,2,3))
a: Test = Test(WrappedArray(1, 2, 3))

scala> val b = Test(Array(1,2,3))
b: Test = Test(WrappedArray(1, 2, 3))

scala> a == b
res0: Boolean = true

scala> a.xs(0) = 0
<console>:13: error: value update is not a member of IndexedSeq[Int]
       a.xs(0) = 0
         ^
```

Personally I'd prefer this approach, letting people access the immutable field directly for filtering and enumerating, rather than providing half a dozen delegating methods.

The other thing was to not inherit `Iterable` since it wraps 2 collections and there's no obvious reason to favor one over the other.